core/src/main/scala/org/http4s/HttpVersion.scala

```diff
@@ -35,20 +35,21 @@ object HttpVersion {
       case "HTTP/1.1" => Either.right(`HTTP/1.1`)
       case "HTTP/1.0" => Either.right(`HTTP/1.0`)
       case _ =>
-        new Parser(s).HttpVersion.run()(Parser.DeliveryScheme.Either).leftMap { _ =>
-          ParseFailure("Invalid HTTP version", s"$s was not found to be a valid HTTP version")
+        parser.parseAll(s).leftMap { case e =>
+          ParseFailure("Invalid HTTP version", e.toString)
         }
     }
 
-  private class Parser(val input: ParserInput)
-      extends org.http4s.internal.parboiled2.Parser
-      with Rfc2616BasicRules {
-    def HttpVersion: Rule1[org.http4s.HttpVersion] =
-      rule {
-        "HTTP/" ~ capture(Digit) ~ "." ~ capture(Digit) ~> { (major: String, minor: String) =>
-          new HttpVersion(major.toInt, minor.toInt)
-        }
-      }
+  private val parser: Parser1[HttpVersion] = {
+    // HTTP-name = %x48.54.54.50 ; HTTP
+    val httpName = P.string1("HTTP")
+
+    // HTTP-version = HTTP-name "/" DIGIT "." DIGIT
+    val httpVersion = httpName ~ P.char('/') *> digit ~ (P.char('.') *> digit)
```

Comment:
Does `httpName ~ P.char('/')` get optimized to `P.string("HTTP/")`?  I'm trying to mimic the RFC, but I'd rather save a few nanoseconds.  Doesn't matter here, but it will start to matter in more complicated parsers.

Reply:
No those are not optimized like that because they are not the same.

If http parses then you don't find the / that is an arresting failure. Strings however peek and automatically backtrack.

And equivalent thing would be `http.soft ~ slash`, this makes a softProduct where if there are no arresting failures it will backtrack.

Lastly, that equivalent thing is not optimized but it should (though I don't know how often it really comes up).

Reply:
Oh, right, that makes sense.

It's not _that_ often that the HTTP ABNF smooshes together two constants.  It's more an academic exercise in how literally we can translate the specs without sacrificing any performance.